VOL refactor and cleanup

qkoziol opened the pull request:
Cleanup and prepare for thread-safety changes.

Big ideas:
- Wrap H5VL_class_t with H5VL_connector_t, so use of the class can be refcounted within the H5VL package, instead of relying on storing an ID within the H5VL_t struct and incrementing & decrementing the ID's refcount.
- Register H5VL_connector_t* for VOL connector IDs, instead of the H5VL_class_t*
- Stop other packages from rummaging around inside H5VL_connector_t and H5VL_object_t data structures, so that the H5VL package can change implementation details without coupled changes throughout the library

Small things:
- Simplified the coding for creating links
- Moved some routines into more logical locations

qkoziol committed:
Rename H5VL_t to H5VL_connector_t

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Switch parameter to H5VL__register_connector

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Change name of routines that operate on connector property objects

So they don't sound like they are operating on connectors.

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Refactor connector object create & free into routines

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Use new connector free routine for freeing new connector object

Also, small code tidying

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Track list of active connectors with global linked list

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Switch to searching list of active VOL connectors, instead of IDs

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

qkoziol committed:
Refactor VOL data structures to remove internal use of IDs

Also simplify and enforce some modularity boundaries around them, so that
other packages can't modify VOL data structures.

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

github-actions[bot] committed:
Committing clang-format changes

qkoziol committed:
Spelling

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Fix link errors in Java wrapper

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol commented on fortran/src/H5VLff.F90:
```diff
@@ -401,6 +401,41 @@ END FUNCTION H5VLunregister_connector
+!!
+!! See C API: @ref H5VLcmp_connector_cls()
+!!
+  SUBROUTINE H5VLcmp_connector_cls_f(are_same, conn_id1, conn_id2, hdferr)
```
This routine is necessary because it's possible to have VOL connector IDs that are different, but point to the same underlying VOL connector class.

qkoziol commented on fortran/test/vol_connector.F90:
```diff
@@ -189,22 +190,30 @@ SUBROUTINE test_registration_by_fapl(total_error)
        CALL H5Pget_vol_id_f(fapl_id, vol_id_out, error)
        CALL check("H5Pget_vol_id_f",error,total_error)
-       CALL VERIFY("H5Pget_vol_id_f", vol_id_out, vol_id, total_error)
+       are_same = .FALSE.
```
Use new routine instead of directly comparing IDs.

qkoziol commented on java/src/hdf/hdf5lib/H5.java:
```diff
@@ -15497,6 +15497,24 @@ public synchronized static native String H5VLget_connector_name(long object_id)
+    /**
+     * @ingroup JH5VL
+     *
+     * H5VLcmp_connector_cls Determines whether two connector identifiers refer to the same connector.
```
This routine is necessary because it's possible to have VOL connector IDs that are different, but point to the same underlying VOL connector class.

qkoziol commented on java/test/TestH5VL.java:
```diff
@@ -98,7 +98,8 @@ public void testH5VLget_connector_id()
             if (connector == null)
-                assertEquals(HDF5Constants.H5VL_NATIVE, native_id);
+                assertTrue("H5.H5VLcmp_connector_cls(H5VL_NATIVE_NAME, native_id)",
+                           H5.H5VLcmp_connector_cls(HDF5Constants.H5VL_NATIVE, native_id));
```
Use new routine instead of directly comparing IDs.

qkoziol commented on src/H5A.c:
```diff
@@ -125,7 +125,7 @@ H5A__create_common(H5VL_object_t *vol_obj, H5VL_loc_params_t *loc_params, const
 
     /* Register the new attribute and get an ID for it */
-    if ((ret_value = H5VL_register(H5I_ATTR, attr, vol_obj->connector, true)) < 0)
+    if ((ret_value = H5VL_register(H5I_ATTR, attr, H5VL_OBJ_CONNECTOR(vol_obj), true)) < 0)
```
These changes, and similar ones in non-H5VL packages, are to remove the code coupling to the H5VL package, allowing it to change implementation of H5VL data structures without rippling through the library.

qkoziol commented on src/H5CX.c:
```diff
@@ -931,33 +931,27 @@ H5CX_retrieve_state(H5CX_state_t **api_state)
 
     /* Keep a copy of the VOL connector property, if there is one */
-    if ((*head)->ctx.vol_connector_prop_valid && (*head)->ctx.vol_connector_prop.connector_id > 0) {
+    if ((*head)->ctx.vol_connector_prop_valid && (*head)->ctx.vol_connector_prop.connector) {
```
VOL connector property stores a pointer to the connector struct now, instead of nesting an ID inside the property list.

qkoziol commented on src/H5F.c:
```diff
@@ -88,12 +88,6 @@ static herr_t H5F__flush_api_common(hid_t object_id, H5F_scope_t scope, void **t
 /* Local Variables */
 /*******************/
 
-/* Declare a free list to manage the H5VL_t struct */
```
Unused (and should not be used outside the H5VL package)

qkoziol commented on src/H5Fpkg.h:
```diff
@@ -311,9 +311,8 @@ struct H5F_shared_t {
 
     /* Cached VOL connector ID & info */
-    hid_t               vol_id;   /* ID of VOL connector for the container */
-    const H5VL_class_t *vol_cls;  /* Pointer to VOL connector class for the container */
```
Unused

qkoziol commented on src/H5Gdeprec.c:
```diff
@@ -312,31 +312,27 @@ H5Glink(hid_t cur_loc_id, H5G_link_t type, const char *cur_name, const char *new
 
         /* Set up new location struct */
         new_loc_params.type                         = H5VL_OBJECT_BY_NAME;
+        new_loc_params.obj_type                     = H5I_get_type(cur_loc_id);
```
Was missing (but worked 🤷‍♂️)

qkoziol commented on src/H5L.c:
```diff
@@ -157,31 +156,19 @@ H5Lmove(hid_t src_loc_id, const char *src_name, hid_t dst_loc_id, const char *ds
                         "Objects are accessed through different VOL connectors and can't be linked");
     }
 
-    /* Construct a temporary source VOL object */
```
Simplified

qkoziol commented on src/H5L.c:
```diff
@@ -271,31 +257,19 @@ H5Lcopy(hid_t src_loc_id, const char *src_name, hid_t dst_loc_id, const char *ds
                         "Objects are accessed through different VOL connectors and can't be linked");
     } /* end if */
 
-    /* Construct a temporary source VOL object */
```
Simplified

qkoziol commented on src/H5L.c:
```diff
@@ -486,12 +456,6 @@ H5L__create_hard_api_common(hid_t cur_loc_id, const char *cur_name, hid_t link_l
     if (H5CX_set_apl(&lapl_id, H5P_CLS_LACC, cur_loc_id, true) < 0)
         HGOTO_ERROR(H5E_LINK, H5E_CANTSET, FAIL, "can't set access property list info");
 
-    /* Set up new location struct */
```
Moved down, after check for H5L_SAME_LOC

qkoziol commented on src/H5L.c:
```diff
@@ -503,45 +467,41 @@ H5L__create_hard_api_common(hid_t cur_loc_id, const char *cur_name, hid_t link_l
                         "Objects are accessed through different VOL connectors and can't be linked");
     } /* end if */
 
-    /* Construct a temporary VOL object */
```
Simplified

qkoziol commented on src/H5O.c:
```diff
@@ -855,29 +854,26 @@ H5Olink(hid_t obj_id, hid_t new_loc_id, const char *new_name, hid_t lcpl_id, hid
                         "Objects are accessed through different VOL connectors and can't be linked");
     } /* end if */
 
-    /* Construct a temporary VOL object */
```
Simplified

qkoziol commented on src/H5Pfapl.c:
```diff
@@ -6152,48 +6144,25 @@ H5P__facc_vol_copy(const char H5_ATTR_UNUSED *name, size_t H5_ATTR_UNUSED size,
     assert(size == sizeof(H5VL_connector_prop_t));
 
-    /* Compare connectors */
-    if (NULL == (cls1 = (H5VL_class_t *)H5I_object(info1->connector_id)))
```
Moved comparison logic into H5VL package

qkoziol commented on src/H5Tcommit.c:
```diff
@@ -77,12 +77,6 @@ static herr_t H5T_destruct_datatype(void *datatype, H5VL_t *vol_connector);
 /*******************/
 
-/* Declare a free list to manage the H5VL_t struct */
-H5FL_EXTERN(H5VL_t);
```
Unused (and shouldn't be here)

qkoziol commented on src/H5VL.c:
```diff
@@ -35,7 +35,7 @@
 #include "H5VLpkg.h"    /* Virtual Object Layer                 */
 
 /* VOL connectors */
-#include "H5VLnative.h" /* Native VOL connector                 */
```
Shouldn't be using public symbols within the library.

qkoziol commented on src/H5VL.c:
```diff
@@ -92,10 +93,20 @@ H5VLregister_connector(const H5VL_class_t *cls, hid_t vipl_id)
+        HGOTO_ERROR(H5E_VOL, H5E_CANTREGISTER, H5I_INVALID_HID, "unable to register VOL class");
+
+    /* Get ID for connector */
+    if ((ret_value = H5I_register(H5I_VOL, connector, true)) < 0)
```
Generally speaking, moved working with IDs out to the API routines.  (There are a few internal ones, but no IDs are stored internally now)

qkoziol commented on src/H5VL.c:
```diff
@@ -314,77 +363,28 @@ H5VLget_connector_id_by_name(const char *name)
+        HGOTO_ERROR(H5E_VOL, H5E_CANTGET, H5I_INVALID_HID, "can't get VOL connector");
 
-/*-------------------------------------------------------------------------
- * Function:    H5VLpeek_connector_id_by_name
```
These "peek" routines probably shouldn't have been made, since we shouldn't have exposed the internal ID publicly.  They have to go now, since there's no more ID.

qkoziol commented on src/H5VLcallback.c:
```diff
@@ -67,6 +66,12 @@ typedef herr_t (*H5VL_reg_opt_oper_t)(void *obj, const H5VL_class_t *cls, H5VL_o
 /* Local Prototypes */
 /********************/
+/* Helper routines */
+static herr_t H5VL__common_optional_op(hid_t id, H5I_type_t id_type, H5VL_reg_opt_oper_t reg_opt_op,
```
Prototype was missing

qkoziol commented on src/H5VLcallback.c:
```diff
@@ -639,6 +645,40 @@ H5VLconnector_info_to_str(const void *info, hid_t connector_id, char **str)
     FUNC_LEAVE_API_NOINIT(ret_value)
 } /* H5VLconnector_info_to_str() */
 
+/*-------------------------------------------------------------------------
```
Moved here, from H5VLint.c, since it matches the other similar calls.

qkoziol commented on src/H5VLcallback.c:
```diff
@@ -677,65 +722,27 @@ H5VLconnector_str_to_info(const char *str, hid_t connector_id, void **info /*out
     FUNC_LEAVE_API_NOINIT(ret_value)
 } /* H5VLget_object */
 
-/*-------------------------------------------------------------------------
```
Simplified out

qkoziol commented on src/H5VLcallback.c:
```diff
@@ -882,56 +897,25 @@ H5VL_unwrap_object(const H5VL_class_t *connector, void *obj)
     FUNC_LEAVE_API_NOINIT(ret_value)
 } /* H5VLunwrap_object */
 
-/*-------------------------------------------------------------------------
```
Simplified out

qkoziol commented on src/H5VLcallback.c:
```diff
@@ -3551,27 +3535,21 @@ H5VL__file_open_find_connector_cb(H5PL_type_t plugin_type, const void *plugin_in
-    /* Silence compiler */
-    (void)plugin_type;
-
-    udata->cls = cls;
```
Simplified this routine a bit, and replaced the H5E_clear_stack call with a pause/resume on the error stack.

qkoziol commented on src/H5VLcallback.c:
```diff
@@ -3663,20 +3631,18 @@ H5VL_file_open(H5VL_connector_prop_t *connector_prop, const char *name, unsigned
-                 * error from original file open failure.
+                 * able to open the file, open the file with it.
                  */
-                H5E_clear_stack();
```
Covered by pause/resume of error stack in previous routine.

qkoziol commented on src/H5VLint.c:
```diff
@@ -113,26 +115,29 @@ static herr_t         H5VL__free_vol_wrapper(H5VL_wrap_ctx_t *vol_wrap_ctx);
+    H5I_VOL,                       /* ID class value */
+    0,                             /* Class flags */
+    0,                             /* # of reserved IDs for class */
+    (H5I_free_t)H5VL__conn_free_id /* Callback routine for closing objects of this class */
```
IDs use H5VL_connector_t now.

qkoziol commented on src/H5VLint.c:
```diff
@@ -113,26 +115,29 @@ static herr_t         H5VL__free_vol_wrapper(H5VL_wrap_ctx_t *vol_wrap_ctx);
 H5FL_DEFINE_STATIC(H5VL_wrap_ctx_t);
 
+/* List of currently active VOL connectors */
+static H5VL_connector_t *H5VL_conn_list_head_g = NULL;
```
Keeps a searchable list of the connectors that have been registered.  Should be very short, so not worried about linear search time for the couple of H5VL API routines that search the list.

qkoziol commented on src/H5VLint.c:
```diff
@@ -305,43 +310,6 @@ H5VL__free_cls(H5VL_class_t *cls, void H5_ATTR_UNUSED **request)
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5VL__free_cls() */
 
-/*-------------------------------------------------------------------------
```
Unnecessary now

qkoziol commented on src/H5VLint.c:
```diff
@@ -638,44 +601,104 @@ H5VL_conn_copy(H5VL_connector_prop_t *connector_prop)
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
```
Connector property comparison routine, used in H5Pfapl.c

qkoziol commented on src/H5VLint.c:
```diff
@@ -759,189 +782,159 @@ H5VL_register_using_existing_id(H5I_type_t type, void *object, H5VL_t *vol_conne
-H5VL_t *
-H5VL_new_connector(hid_t connector_id)
+H5VL_object_t *
+H5VL_create_object(void *object, H5VL_connector_t *vol_connector)
```
Better name, simpler code, now.

qkoziol replied:
Moved up here, during some small code rearrangement.  Unchanged.

qkoziol commented on src/H5VLint.c:
```diff
@@ -759,189 +782,159 @@ H5VL_register_using_existing_id(H5I_type_t type, void *object, H5VL_t *vol_conne
  *-------------------------------------------------------------------------
  */
-hid_t
-H5VL_register_using_vol_id(H5I_type_t type, void *obj, hid_t connector_id, bool app_ref)
```
H5VL_register_using_vol_id is only used in the testing code now, so I moved it to H5VLtest.c

qkoziol commented on src/H5VLint.c:
```diff
@@ -759,189 +782,159 @@ H5VL_register_using_existing_id(H5I_type_t type, void *object, H5VL_t *vol_conne
-hid_t
-H5VL_register_using_vol_id(H5I_type_t type, void *obj, hid_t connector_id, bool app_ref)
+static H5VL_connector_t *
+H5VL__conn_create(H5VL_class_t *cls)
```
This used to be H5VL_new_connector

qkoziol commented on src/H5VLint.c:
```diff
@@ -759,189 +782,159 @@ H5VL_register_using_existing_id(H5I_type_t type, void *object, H5VL_t *vol_conne
  *-------------------------------------------------------------------------
  */
-H5VL_object_t *
-H5VL_create_object_using_vol_id(H5I_type_t type, void *obj, hid_t connector_id)
```
Deleted

qkoziol commented on src/H5VLint.c:
```diff
@@ -759,189 +782,159 @@ H5VL_register_using_existing_id(H5I_type_t type, void *object, H5VL_t *vol_conne
-H5VL_object_t *
-H5VL_create_object_using_vol_id(H5I_type_t type, void *obj, hid_t connector_id)
+static herr_t
+H5VL__conn_find(H5PL_vol_key_t *key, H5VL_connector_t **connector)
```
Searches the list of registered connectors

qkoziol commented on src/H5VLint.c:
```diff
@@ -992,23 +986,123 @@ H5VL_conn_dec_rc(H5VL_t *connector)
+ *-------------------------------------------------------------------------
+ */
+htri_t
+H5VL_conn_same_class(const H5VL_connector_t *conn1, const H5VL_connector_t *conn2)
```
For other packages in the library to use.

qkoziol commented on src/H5VLint.c:
```diff
@@ -1454,268 +1543,132 @@ H5VL__is_connector_registered_by_name(const char *name)
  *-------------------------------------------------------------------------
  */
-hid_t
-H5VL__get_connector_id(hid_t obj_id, bool is_api)
```
Unused now

qkoziol commented on src/H5VLint.c:
```diff
@@ -1454,268 +1543,132 @@ H5VL__is_connector_registered_by_name(const char *name)
-    FUNC_LEAVE_NOAPI(ret_value)
-} /* end H5VL__peek_connector_id_by_value() */
-
-/*-------------------------------------------------------------------------
```
Moved to H5VLcallback.c

qkoziol commented on src/H5VLnative.c:
```diff
@@ -186,26 +186,26 @@ static const H5VL_class_t H5VL_native_cls_g = {
  *-------------------------------------------------------------------------
  */
-hid_t
+H5VL_connector_t *
```
Refactored since H5VL package doesn't store IDs any longer

qkoziol commented on src/H5VLnative.h:
```diff
@@ -25,9 +25,6 @@
 /* Public Macros */
 /*****************/
 
-/* Identifier for the native VOL connector */
```
Move private pieces to private header

qkoziol commented on src/H5VLpkg.h:
```diff
@@ -36,37 +36,51 @@
 /****************************/
 
+/* Internal struct to track VOL connectors */
+struct H5VL_connector_t {
```
One connector object per VOL connector class registered with the library

qkoziol commented on src/H5VLprivate.h:
```diff
@@ -27,28 +27,33 @@
 
+/* If the module using this macro is allowed access to the private variables, access them directly */
+#ifdef H5VL_MODULE
+#define H5VL_OBJ_RC(VOL_OBJ)         ((VOL_OBJ)->rc)
```
Getters, for the rest of the library packages.  (Similar idiom as in H5Fprivate.h)

qkoziol commented on test/event_set.c:
```diff
@@ -395,7 +396,12 @@ test_es_get_requests(void)
         TEST_ERROR;
-    if (connector_ids[0] != connector_ids_g[0])
+    cmp_value = 0;
+    if (H5VLcmp_connector_cls(&cmp_value, connector_ids[0], connector_ids_g[0]) < 0)
```
Use new comparison routine in tests.

qkoziol commented:
@mattjala - this is the PR that we talked about a week and a half ago.  Should replace https://github.com/HDFGroup/hdf5/pull/4635 and fix https://github.com/HDFGroup/hdf5/issues/4634

mattjala commented on src/H5ES.c:
```diff
@@ -122,10 +122,8 @@ H5ESinsert_request(hid_t es_id, hid_t connector_id, void *request)
-    /* Create new VOL connector object, using the connector ID */
-    if (NULL == (connector = H5VL_new_connector(connector_id)))
-        HGOTO_ERROR(H5E_EVENTSET, H5E_CANTCREATE, FAIL, "can't create VOL connector object");
+    if (NULL == (connector = H5I_object_verify(connector_id, H5I_VOL)))
```
Since this routine doesn't create a connector anymore, the later call to `H5VL_conn_dec_rc()` should be removed. Any rc cleanup on the connector will be handled by the VOL object cleanup.

qkoziol replied:
Good catch, fixed in upcoming commit

mattjala commented on src/H5Fprivate.h:
```diff
@@ -163,7 +162,6 @@ typedef struct H5F_t H5F_t;
 #define H5F_NULL_FSM_ADDR(F)           (H5F_get_null_fsm_addr(F))
 #define H5F_GET_MIN_DSET_OHDR(F)       (H5F_get_min_dset_ohdr(F))
 #define H5F_SET_MIN_DSET_OHDR(F, V)    (H5F_set_min_dset_ohdr((F), (V)))
-#define H5F_VOL_CLS(F)                 (H5F_get_vol_cls(F))
```
Should these be replaced with `H5F_VOL_CONN` macros? They wouldn't be used anywhere at present, but it seems like it would be more consistent.

qkoziol replied:
Refactoring the H5F package is out of my scope for this PR, but I'll see about making the change in another in the future.

derobins commented on src/H5VLconnector.h:
```diff
@@ -1105,14 +1105,6 @@ H5_DLL void *H5VLobject(hid_t obj_id);
-/**
- * \ingroup H5VLDEV
- */
-H5_DLL hid_t H5VLpeek_connector_id_by_value(H5VL_class_value_t value);
```
This definitely needs a RELEASE.txt note

qkoziol replied:
Yes, will be in next PR, coming shortly.

mattjala commented on src/H5L.c:
```diff
@@ -503,45 +467,41 @@ H5L__create_hard_api_common(hid_t cur_loc_id, const char *cur_name, hid_t link_l
 
+    /* Set the connector to use for async operations */
+    if (connector)
+        *connector = (link_vol_obj ? H5VL_OBJ_CONNECTOR(link_vol_obj) : H5VL_OBJ_CONNECTOR(curr_vol_obj));
```
It looks like this has a minor inconsistency in behavior - the old version checks curr_vol_obj for a connector before falling back to the connector on link_vol_obj, but this switches the order.

qkoziol replied:
The logic for this line matches the logic for the H5VL_object_t that is passed in to H5VL_link_create.   The older logic for which object to use resolved down to what I have now, or equivalent.

mattjala commented on src/H5L.c:
```diff
@@ -503,45 +467,41 @@ H5L__create_hard_api_common(hid_t cur_loc_id, const char *cur_name, hid_t link_l
 
+    /* Set the connector to use for async operations */
+    if (connector)
+        *connector = (link_vol_obj ? H5VL_OBJ_CONNECTOR(link_vol_obj) : H5VL_OBJ_CONNECTOR(curr_vol_obj));
```
Since this returns a pointer to the connector, I think it should increment the ref count and require the caller to release it after it's done.

qkoziol replied:
It's strictly local within the source module and only used in one spot, so I don't think that is necessary here.

mattjala commented on src/H5Pfapl.c:
```diff
@@ -5748,20 +5748,20 @@ H5Pget_page_buffer_size(hid_t plist_id, size_t *buf_size /*out*/, unsigned *min_
-    if (NULL == H5I_object_verify(vol_id, H5I_VOL))
-        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a VOL connector ID");
+    /* Sanity check */
+    assert(connector);
```
I think there's a case from `H5VL__set_def_conn()` where this function could be called with a NULL connector - if the `connector_is_registered = H5VL__is_connector_registered_by_name(tok)` succeeds and the `connector` assigned to `H5VL_def_conn_s` is NULL.

qkoziol replied:
I don't see any code path in H5VL__set_def_conn where connector could be NULL.

mattjala replied:
I think you're correct; I read the function wrong.

mattjala commented on src/H5VL.c:
```diff
@@ -314,77 +363,28 @@ H5VLget_connector_id_by_name(const char *name)
+        HGOTO_ERROR(H5E_VOL, H5E_CANTGET, H5I_INVALID_HID, "can't get VOL connector");
 
-/*-------------------------------------------------------------------------
- * Function:    H5VLpeek_connector_id_by_name
```
If these are removed, the relevant sections in `VOLConnGuide.dox` should be deleted.

qkoziol replied:
Yes, will be in the next PR, now that we've discussed it.

mattjala commented on src/H5VL.c:
```diff
@@ -314,77 +363,28 @@ H5VLget_connector_id_by_name(const char *name)
+        HGOTO_ERROR(H5E_VOL, H5E_CANTGET, H5I_INVALID_HID, "can't get VOL connector");
 
-/*-------------------------------------------------------------------------
- * Function:    H5VLpeek_connector_id_by_name
```
What exactly is the distinction between `H5VLget_connector_id(_by_name/value)` and `H5VLpeek_connector_id_by_name/value` such that the peek routines should be removed? All of them now return the ID to a H5VL_connector_t instead of a H5VL_class_t.

 If it has to do with the expectation that a connector can access its class under the ID, couldn't that still be achieved by looking at the `cls` field of the connector object?

qkoziol replied:
The peek routines were returning an internal ID from the library's data structure and callers were told "don't close this ID", which is a problem in several ways.  The existing get routines behave like all the other get-an-ID routines in the library's API.

mattjala commented on src/H5VL.c:
```diff
@@ -674,25 +675,28 @@ H5VLobject_is_native(hid_t obj_id, hbool_t *is_native)
     /* Check args */
     if (!file_obj)
         HGOTO_ERROR(H5E_ARGS, H5E_UNINITIALIZED, FAIL, "no file object supplied");
-    if (NULL == (dtype = (H5T_t *)H5I_object_verify(dtype_id, H5I_DATATYPE)))
```
Why remove this cast?

qkoziol replied:
They are unnecessary noise.

mattjala commented on src/H5VLcallback.c:
```diff
@@ -5909,20 +5872,20 @@ herr_t
+    if (NULL == (connector = H5I_object_verify(connector_id, H5I_VOL)))
         HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a VOL connector ID");
 
-    /* Bypass the H5VLint layer, calling the VOL callback directly */
```
What was the original reason to bypass the H5VLint layer here?

qkoziol replied:
Not certain 🤷‍♂️

qkoziol replied:
Better to have it be consistent, in the long term

qkoziol committed:
Address review comments

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

mattjala commented on src/H5VLint.c:
```diff
@@ -1454,268 +1543,132 @@ H5VL__is_connector_registered_by_name(const char *name)
 done:
     FUNC_LEAVE_NOAPI(ret_value)
-} /* end H5VL__get_connector_id_by_value() */
+} /* end H5VL__get_connector_id_by_name() */
```
`H5VL__get_connector_id_by_name()` -> `H5VL__get_connector_by_name()`

qkoziol committed:
Fix comment

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

mattjala commented on tools/lib/h5tools.c:
```diff
@@ -671,12 +671,8 @@ h5tools_set_fapl_vol(hid_t fapl_id, h5tools_vol_info_t *vol_info)
                 /* Check for VOL connectors that ship with the library, then try
                  * registering by name if that fails.
                  */
-                if (!strcmp(vol_info->u.name, H5VL_NATIVE_NAME)) {
```
This looks like it would cause problems when trying to use h5tools with the native VOL

qkoziol replied:
Tools use the native connector by default; no command-line option is necessary.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

jhendersonHDF commented:
General comment: `H5VLpeek_connector_id_by_name` is used by the log-based VOL and `H5VLpeek_connector_id_by_value` is used by the REST and DAOS VOLs, so we'll want to decide how to handle that switchover gracefully.

jhendersonHDF commented on release_docs/RELEASE.txt:
```diff
@@ -877,6 +888,7 @@ Bug Fixes since HDF5-1.14.0 release
 
       Fixes GitHub #4656 and GitHub #4662
 
+>>>>>>> origin/develop
```
merge artifact

qkoziol replied:
Fixed, thanks!

qkoziol commented:
> General comment: `H5VLpeek_connector_id_by_name` is used by the log-based VOL and `H5VLpeek_connector_id_by_value` is used by the REST and DAOS VOLs, so we'll want to decide how to handle that switchover gracefully.

Yes.  I added info to the RELEASE.txt note suggesting the correct action for developers who are using them.   Anything else needed?   We can update the connectors after the merge.

qkoziol committed:
Remove merge trash

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

jhendersonHDF commented on java/src/jni/h5Constants.c:
```diff
@@ -3500,7 +3500,7 @@ Java_hdf_hdf5lib_HDF5Constants_H5VL_1CAP_1FLAG_1THREADSAFE(JNIEnv *env, jclass c
 JNIEXPORT jlong JNICALL
 Java_hdf_hdf5lib_HDF5Constants_H5VL_1NATIVE(JNIEnv *env, jclass cls)
 {
-    return H5VL_NATIVE;
```
Doesn't this change the semantics for the `H5VL_NATIVE` constant in that previously it could be used and forgotten, whereas now you have to close the ID given back?

qkoziol replied:
Hmm, yes, that's correct and unintended.  I'll fix this back to a global hid_t in the C library and then propagate the change out to Java & FORTRAN.

qkoziol replied:
Fixed in my commit just now.

Bonus for the cleanup on the internal passthru connector: found and removed another use of a public API call from within the library.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

qkoziol committed:
Switch VOL connector macros to use "H5OPEN, <id>" idiom

Aligns them with the way the H5T, H5P, H5E packages return an ID back to
an application, and removes calling an internal routine from the public header.

Extra work for the internal passthru connector, to keep the main passthru
source file using only public API calls.

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

github-actions[bot] committed:
Committing clang-format changes

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

qkoziol committed:
Fixe speeling eror

Signed-off-by: Quincey Koziol <quincey@koziol.cc>

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup

qkoziol commented:
@jhendersonHDF - Any other comments for this PR?   I'd like to get it merged ASAP after the 1.16 branch is created, so I can queue the next large one up.

jhendersonHDF commented:
@qkoziol I don't think I have any other complaints in particular, though I believe @fortnern mentioned he still needed to review it and I also believe we decided to discuss this PR at the next working group meeting.

qkoziol commented:
I've already discussed it at last week's meeting, although I'm happy to run through it again.  I believe it's ready for any final review and merge.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into vol_cleanup
